core/src/database/PostgreSQLBackend.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ *
+ * PostgreSQLBackend
+ *
+ * Created by El Khalil Bellakrid on 06/12/2019.
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include "PostgreSQLBackend.h"
+#include <utils/Exception.hpp>
+#include <api/ConfigurationDefaults.hpp>
+#ifdef PG_SUPPORT
+    #include <soci-postgresql.h>
+#endif
+using namespace soci;
+
+namespace ledger {
+    namespace core {
+        PostgreSQLBackend::PostgreSQLBackend() : DatabaseBackend(),
+                                                 _connectionPoolSize(api::ConfigurationDefaults::DEFAULT_PG_CONNECTION_POOL_SIZE)
+        {}
+
+        PostgreSQLBackend::PostgreSQLBackend(int32_t connectionPoolSize) : DatabaseBackend(),
+                                                                           _connectionPoolSize(connectionPoolSize)
+        {}
+
+        int32_t PostgreSQLBackend::getConnectionPoolSize() {
+            return _connectionPoolSize;
+        }
+
+        void PostgreSQLBackend::init(const std::shared_ptr<ledger::core::api::PathResolver> &resolver,
+                                  const std::string &dbName,
+                                  const std::string &password,
+                                  soci::session &session) {
+            _dbName = dbName;
+            setPassword(password, session);
+        }
+
+        void PostgreSQLBackend::setPassword(const std::string &password,
+                                         soci::session &session) {
+            if (_dbName.empty()) {
+                throw make_exception(api::ErrorCode::DATABASE_EXCEPTION, "Database should be initiated before setting password.");
+            }
+            session.close();
+#ifdef PG_SUPPORT
```

Comment:
I don’t get why you need this. That file should be part of a compilation unit that is conditionally checked against `PG_SUPPORT` so that you don’t have to make that kind of test inside your code.

Reply:
This file is always part of compilation, even if I exclude it from compilation on the `CMakeLists.txt` level, I'll have to do same thing with `PG_SUPPORT` on `DatabaseBackend` side, so in the end it introduces more complications than it solves ! 
WDYT ?

Reply:
I mean, you’re in a file `PostgreSQLBackend`, so that file should always be compiled with `PG_SUPPORT`. Compiling that file without `PG_SUPPORT` doesn’t make sense (or maybe it does but then the filename should be something else).

My point is:

- Instead of always compiling that file and check for `PG_SUPPORT` inside the file…
- Compile the file only when `PG_SUPORT` is set and then you have zero assumption to make inside the implementation of the file.

Reply:
However, maybe that is a bit boring to do or you don’t want to (it would mean removing GLOB patterns in a `CMakeFiles.txt` and add the source conditionally). I’ll let you decide what you prefer to do, even though I think what I describe is a bit cleaner.

Reply:
I got your previous point, but even by compiling that only with `PG_SUPPORT` you'll have to do the check on `DatabaseBackend` because we include `PostgreSQLBackend` (that was my previous comment)